### ISSUE TITLE
Constant crashes when attempting to draw on canvas since upgrade to GNOME 45

### ISSUE THREAD

**Issue by fotnite-vevo** [flxzt/rnote#1061]
**Describe the bug**  
Nearly any action (attempting to draw on the canvas, attempting to move copied in photo) will potentially crash the app, and occasionally crash GNOME Shell. Only occurs when running app under Wayland (never crashes under XWayland)

**To Reproduce**  
1. Attempt to draw on canvas, attempt to move object, etc.

**Expected behavior**  
Rnote doesn't crash

**Console Output**  
```
2024-03-29T23:05:52.845016Z DEBUG rnote: .. tracing subscriber initialized.
rnote: ../cairo/src/cairo-pattern.c:3831: _cairo_pattern_get_extents: Assertion `status == CAIRO_STATUS_SUCCESS' failed.
fish: Job 1, 'RUST_LOG=rnote=debug RUST_BACKT…' terminated by signal SIGABRT (Abort)
```

**Desktop (please complete the following information):**  
 - OS: Arch Linux
 - App Version: Rnote v0.10.1 (also applies to prior versions)
 - Installation Source: Arch Linux Community
 - Desktop Environment: GNOME 46.0
 - Display Server: Wayland
 - Input Source: Mouse and Keyboard, Tablet

**Comment by fotnite-vevo** [comment]
Alright, I did a bit more testing, and it seems to be **related to fractional scaling.** I usually have fractional scaling set to 1.25x, I still get crashes on 1.5x, but not on 1x, 1.75x, or 2x. but Not sure if it's an upstream bug or not but it seems like it might be.

**Comment by Doublonmousse** [comment]
Gnome 46 means running on newer gtk (4.14) and libadwaita (1.5) versions.

gtk 4.14 made `ngl` the default renderer and that seems like the most likely suspect (see https://gitlab.gnome.org/GNOME/gtk/-/issues/?sort=created_date&state=opened&search=ngl&first_page_size=20)

Maybe follow https://gitlab.gnome.org/GNOME/gtk/-/issues/6574#note_2063669 and set `GSK_RENDERER=gl` or to another renderer to test if ngl is the issue (and report issues upstream to the gtk gitlab).

It will be a while before we can seriously look at gtk 4.14 though (probably need to wait until the new versions are available on all platforms
- mac : brew, up to date
- windows : msys, not up to date but seems close
- linux : I'm not sure flatpak runtimes are up to date, debian packages are not on gtk 4.14, and for fedora/ubuntu-based things the latest versions will probably come not sooner than the next release (fedora 40 and next ubuntu lts)

Once this is available we'll be able to look at gtk 4.14 and related issues (and look back on any previous gtk issues and workarounds)

**Comment by fotnite-vevo** [comment]
> Gnome 46 means running on newer gtk (4.14) and libadwaita (1.5) versions.
> 
> gtk 4.14 made `ngl` the default renderer and that seems like the most likely suspect (see https://gitlab.gnome.org/GNOME/gtk/-/issues/?sort=created_date&state=opened&search=ngl&first_page_size=20)
> 
> Maybe follow https://gitlab.gnome.org/GNOME/gtk/-/issues/6574#note_2063669 and set `GSK_RENDERER=gl` or to another renderer to test if ngl is the issue (and report issues upstream to the gtk gitlab).
> 
> It will be a while before we can seriously look at gtk 4.14 though (probably need to wait until the new versions are available on all platforms
> 
>     * mac : brew, up to date
> 
>     * windows : msys, not up to date but seems close
> 
>     * linux : I'm not sure flatpak runtimes are up to date, debian packages are not on gtk 4.14, and for fedora/ubuntu-based things the latest versions will probably come not sooner than the next release (fedora 40 and next ubuntu lts)
> 
> 
> Once this is available we'll be able to look at gtk 4.14 and related issues (and look back on any previous gtk issues and workarounds)

Yes, it seems like this is the source of the issue. I'll report this upstream when I have the time to do so.

**Comment by UnlimitedKaos** [comment]
>Maybe follow [\[..\]](https://gitlab.gnome.org/GNOME/gtk/-/issues/6574#note_2063669) and set `GSK_RENDERER=gl`

I also experienced crashes with the arch community package (fwiw Im using Plasma6/Wayland), and this seems to fix it

**Comment by safizn** [comment]
Experiencing crashes too when drawing. Probably related to fractional scaling as mentioned above. 

Crashing version: 
```
   Commit: 25f606dcb4150523065a0167c44f074b31288ae0f40d2cc538c5ecd6c3b7f6c8
   Parent: eeb5736a683913857b95ea5fd2dcd8f4cae1acd65aa375cc879623958c9ce9e4
   Subject: update to v0.10.2 (#54) (5f9c3581)
   Date: 2024-04-06 13:24:35 +0000
```

Passing version _(didn't test versions in-between)_: This version does not exhibit the crashing issue. 
```
    Commit: dc829353acef99cde736910ca4bafcf6ed4ecdb9623056b0d48e043ef7c4254e
    Subject: update to `v0.9.4` (#48) (ea975b81)
    Date: 2023-12-13 08:51:52 +0000
```

**Comment by danielwmoore** [comment]
A little extra info, in case it helps.

I'm not using fractional scaling, but am running Ubuntu 24.04 with Gnome 46, and experience the same crashes. Drawing then attempting to zoom or pan often crashes. Drawing and switching to eraser crashes. Takes the entire desktop with it and I have to log back in, having lost any unsaved work in all my apps.

Having the same issue with GIMP crashing when I switch from mouse input to pen input, didn't happen before moving to 24.04 with Gnome 46, so probably the new renderer for sure.

I have not had any crashes with Xournal++ using the pen. rnote and GIMP crash regularly when using the pen. I know, these apps aren't rnote, but I thought it might help in hammering down exactly where the issue is.

**Comment by Doomsdayrs** [comment]
Adding a bit more info.

This issue will eventually cause the entire gnome-shell to crash as well.

**Comment by mbeps** [comment]
Hello,
I also have the same issue. Sometimes the entire app crashes but in more sever cases, the entire Gnome Shell crashes. I am currently on Gnome 46 on Fedora 40 using Wayland. The stylus is some sort of Wacom solution used by Dell. 

`flatpak run com.github.flxzt.rnote
rnote: ../src/cairo-pattern.c:3831: _cairo_pattern_get_extents: Assertion `status == CAIRO_STATUS_SUCCESS' failed.`

**Comment by fotnite-vevo** [comment]
A bit more probing found that this issue does not occur for the Brush tool when path modelling is set to simple. Dragging images can still cause crashes.

**Comment by fotnite-vevo** [comment]
> Hello, I also have the same issue. Sometimes the entire app crashes but in more sever cases, the entire Gnome Shell crashes. I am currently on Gnome 46 on Fedora 40 using Wayland. The stylus is some sort of Wacom solution used by Dell.
> 
> ```
> flatpak run com.github.flxzt.rnote
> rnote: ../src/cairo-pattern.c:3831: _cairo_pattern_get_extents: Assertion `status == CAIRO_STATUS_SUCCESS' failed.
> ```

Since you're running the flatpak version, your shell crashes are likely caused by (this separate issue)[https://gitlab.gnome.org/GNOME/mutter/-/issues/3372], which has already been patched upstream in Mutter but has not yet shipped to distributions.

**Comment by BuonHobo** [comment]
Is it possible to make a new release or at least backport this fix to the flathub version? Unfortunately it's currently unusable for me because I have fractional scaling

**Comment by fotnite-vevo** [comment]
> As the renderer will change to vulkan with the next gtk stable release (4.16) on linux, we can revisit this then (and not lose time searching for nothing).

In light of this, I'd like to point out that on the current flatpak version, the same crash occurs, but with a different console output:

```
Gsk:ERROR:../gsk/gpu/gskvulkanimage.c:260:gsk_vulkan_image_new: assertion failed: (width > 0 && height > 0)
Bail out! Gsk:ERROR:../gsk/gpu/gskvulkanimage.c:260:gsk_vulkan_image_new: assertion failed: (width > 0 && height > 0)
```

Which seems like essentially the same bug, but with the Vulkan renderer instead of with Cairo.

**Comment by flxzt** [comment]
Rnote v0.11.0, which was just released, forces the renderer to 'gl' in response to the many crash reports posted here.

Eventually this should be re-evaluated at a later date, when Gtk's new renderers become more stable.

**Comment by Doublonmousse** [comment]
I've been trying again on gtk 4.16.5 and I can't reproduce the crash (be it with ngl or vulkan). 
So the original issue seems fixed.

Now the vulkan renderer isn't without issues though, at least on my machine.
So I'm not sure if it's time to remove the workaround or not (and if we do, maybe we should test the gtk version before doing so ? I'm not sure of the chances of a new release being run on gtk 4.14 to decide on this).

**Comment by SomeonetookSomeonetookmeep** [comment]
I'm on ver. 11, and getting similar random crashes. I'm on win 10 on a surface pro four, and it freezes a few seconds before it just shuts down, with no apparent affect on the rest of the OS. 

To try to mitigate it I've been cranking up the autosave time

**Comment by matthew-j-hooper** [comment]
+1 for this issue on version 12. Im getting rnote crashes (no crashes in KDE) when attempting to use draw or erase tools. I have downgraded to version 11 of rnote which has fixed the issue. 

Running in the terminal produces this error upon crash:
```
**
Gsk:ERROR:../gsk/gpu/gskvulkanimage.c:271:gsk_vulkan_image_new: assertion failed: (width > 0 && height > 0)
Bail out! Gsk:ERROR:../gsk/gpu/gskvulkanimage.c:271:gsk_vulkan_image_new: assertion failed: (width > 0 && height > 0)
```

- Running rnote through `export GSK_RENDERER="gl"` then `flatpak run com.github.flxzt.rnote` doesnt fix the issue.
- Setting fractional scaling to 175% fixes the issue.
- `gtk4-launch --version` outputs `4.16.13`

Using:
Fedora 41
KDE Plasma 6.3.4
Wayland
Kernel 6.13.10
Fractional scaling 170%

Installed with:
Flatpak

**Comment by Doublonmousse** [comment]
@matthew-j-hooper I'm opening a new issue for the vulkan part (so that we treat one error message at a time per issue)

> Running rnote through export GSK_RENDERER="gl" then flatpak run com.github.flxzt.rnote doesnt fix the issue.
So in that case you get back the 
```
rnote: ../cairo/src/cairo-pattern.c:3831: _cairo_pattern_get_extents: Assertion `status == CAIRO_STATUS_SUCCESS' failed.
```
message ?

The system gtk version doesn't mean too much with flatpaks, if you want the actual gtk version that's used by the flatpak, do
```
flatpak --env=GTK_DEBUG=interactive run com.github.flxzt.rnote
```
then two window open and the info is then here (with the used renderer backend as well) :

 
![Image](https://github.com/user-attachments/assets/75dcf97b-1996-451e-a7fa-ececeeb78356)

**Comment by matthew-j-hooper** [comment]
@Doublonmousse I just checked and can confirm both points you've brought up. When running with GSK_RENDERER="gl" I'll get the error message you've brought up instead of the vulkan error. Also confirming that my GTK version is 4.18.2 using flatpak.

**Comment by Doublonmousse** [comment]
The fix should be included in gtk 4.18.5 (we'll need to wait for gnome 48.2 at the end of this month to release so that the flatpak has the latest gtk version, then I'll be able to do a nightly to confirm this is fixed)

**Comment by Doublonmousse** [comment]
Hello, the GNOME Application Platform was updated to 48.2 for flatpaks and contains the gtk update that fixed this crash (gtk4 updated to 4.18.5).

Be sure to do `flatpak update` and update your `org.gnome.Platform` flatpak and the issue should be fixed.

To be extra sure, you can run 
```
flatpak run --env=GTK_DEBUG=interactive com.github.flxzt.rnote 
```
and check in the second window that the gtk version is 4.18.5

![Image](https://github.com/user-attachments/assets/c5370dd2-0a09-41d1-9481-1f244e5516b7)

**Comment by matthew-j-hooper** [comment]
I can confirm that, for the flatpak version, the 4.18.5 update has fixed this issue for me
